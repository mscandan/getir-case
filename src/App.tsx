import React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react';
import { ThemeProvider } from 'styled-components';
import { appTheme } from 'constants/theme';
import GlobalStyles from 'lib/globalStyles';
import { store, persistor } from 'store';

const App: React.FC = () => {
  return (
    <ReduxProvider store={store}>
      <PersistGate persistor={persistor}>
        <ThemeProvider theme={appTheme}>
          <GlobalStyles />
        </ThemeProvider>
      </PersistGate>
    </ReduxProvider>
  );
};

export default App;
