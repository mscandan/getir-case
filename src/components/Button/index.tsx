import React from 'react';
import styled from 'styled-components';

const StyledButton = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;
  height: 22px;
  cursor: pointer;
  border-radius: 2px;
  font-weight: 600;
  font-size: 12px;
  line-height: 20px;
  background-color: ${({ theme }) => theme.COMPONENT_BG.BASE};
  color: ${({ theme }) => theme.TEXT_LIGHT.HIGHLIGHT};

  &:hover {
    cursor: pointer;
  }
`;

const StyledButtonContentContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const Button: React.FC<React.ButtonHTMLAttributes<HTMLButtonElement> & React.PropsWithChildren> = React.memo(
  ({ children, onClick, type = 'button', ...rest }) => {
    return (
      <StyledButton type={type} onClick={onClick} {...rest} data-testid="button">
        <StyledButtonContentContainer>{children}</StyledButtonContentContainer>
      </StyledButton>
    );
  },
);
