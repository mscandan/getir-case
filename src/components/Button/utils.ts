import { DefaultTheme } from 'styled-components';
import { ButtonSize, ButtonType } from 'types';

export const getBorderStyling = (theme: DefaultTheme) => {
  return `2px solid ${theme.COMPONENT_BG.BASE}`;
};

export const getFontSize = (size: ButtonSize) => {
  switch (size) {
    case 'md':
      return 13;
    case 'lg':
      return 14;
    default:
      return 12;
  }
};

export const getBackgroundColor = (theme: DefaultTheme, buttonType: ButtonType) => {
  switch (buttonType) {
    case 'solid':
      return theme.COMPONENT_BG.BASE;
    default:
      return 'transparent';
  }
};

export const getSize = (size: ButtonSize) => {
  switch (size) {
    case 'lg':
      return '16px 24px';
    case 'md':
      return '16px 6px';
    default:
      return '6px 2px';
  }
};

export const getColor = (theme: DefaultTheme, buttonType: ButtonType) => {
  switch (buttonType) {
    case 'solid':
      return theme.TEXT_LIGHT.HIGHLIGHT;
    case 'outline':
      return theme.COMPONENT_BG.BASE;
    default:
      return theme.TEXT_DARK.BASE;
  }
};
