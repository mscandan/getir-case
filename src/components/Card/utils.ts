import { CardSizeType } from 'types';

export const getCardWidth = (size: CardSizeType) => {
  switch (size) {
    case 'sm':
      return '296px';
    case 'lg':
      return '608px';
    default:
      return '100%';
  }
};
