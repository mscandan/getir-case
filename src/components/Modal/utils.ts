export const isGivenParameterNumber = (parameter: number | string) => typeof parameter === 'number';
