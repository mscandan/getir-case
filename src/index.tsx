import React from 'react';
import ReactDOM from 'react-dom/client';
import { Provider as ReduxProvider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react';
import { store, persistor } from 'store';
import App from './App';

const root = ReactDOM.createRoot(document.getElementById('root') as HTMLElement);
root.render(
  <React.StrictMode>
    <ReduxProvider store={store}>
      <PersistGate persistor={persistor}>
        <App />
      </PersistGate>
    </ReduxProvider>
  </React.StrictMode>,
);
