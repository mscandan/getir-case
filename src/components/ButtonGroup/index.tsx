/* eslint-disable max-len */
import React from 'react';
import styled from 'styled-components';
import { ButtonGroupDataType } from 'types';
import { Button } from '../Button';

interface ButtonGroupProps {
  data: Array<ButtonGroupDataType>; // button's to render on the screen
  onSelectedButtonChange: (newSelectedButton: ButtonGroupDataType) => void; // function to invoke on choosing new selected button
  selectedButtonId?: string; // selected button id, default is given data's [0] element's id
}

const StyledButtonGroup = styled.div`
  display: flex;
  align-items: center;
`;

const StyledButton = styled(Button)<{ isSelected: boolean }>`
  height: 30px;
  padding: 0 16px;
  margin-right: 8px;
  font-size: 13px;
  line-height: 18px;
  background-color: ${({ isSelected, theme }) => (isSelected ? theme.COMPONENT_BG.BASE : theme.COMPONENT_BG.HIGHLIGHT)};
  color: ${({ isSelected, theme }) => (isSelected ? theme.TEXT_LIGHT.HIGHLIGHT : theme.COMPONENT_BG.BASE)};
`;

export const ButtonGroup: React.FC<ButtonGroupProps> = ({
  data,
  onSelectedButtonChange,
  selectedButtonId,
  ...rest
}) => {
  const [activeButtonId, setActiveButtonId] = React.useState(selectedButtonId ?? data[0].id);

  const handleOnClick = (id: string, label: string) => {
    setActiveButtonId(id);
    if (onSelectedButtonChange) onSelectedButtonChange({ id, label });
  };

  return (
    <StyledButtonGroup {...rest}>
      {data.map(button => (
        <StyledButton
          onClick={() => handleOnClick(button.id, button.label)}
          isSelected={activeButtonId === button.id}
          key={button.id}
        >
          {button.label}
        </StyledButton>
      ))}
    </StyledButtonGroup>
  );
};
