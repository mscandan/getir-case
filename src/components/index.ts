import { Button } from './Button';
import { Checkbox } from './Checkbox';
import { Radios } from './Radios';
import { Icon } from './Icon';
import { Card } from './Card';
import { ButtonGroup } from './ButtonGroup';
import { Input } from './Input';
import { Counter } from './Counter';
import { Modal } from './Modal';
import { LoadingSpinner } from './LoadingSpinner';
import { ProductItem } from './ProductItem';
import { Pagination } from './Pagination';

export {
  Button,
  Checkbox,
  Radios,
  Icon,
  Card,
  ButtonGroup,
  Input,
  Counter,
  Modal,
  LoadingSpinner,
  ProductItem,
  Pagination,
};
