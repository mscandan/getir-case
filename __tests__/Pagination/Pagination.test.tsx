import React from 'react';
import renderer from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react';
import { componentRenderer } from '../utils';
import { Pagination } from '../../src/components';

describe('Pagination Tests', () => {
  it('should match snapshot', () => {
    const comp = renderer.create(
      componentRenderer(<Pagination pageCount={10} selectedPageIndex={0} onSelectedPageIndexChange={jest.fn()} />),
    );
    expect(comp).toMatchSnapshot();
  });

  it('should trigger given onChange function on page number click', () => {
    const onChange = jest.fn();
    const { queryAllByRole } = render(
      componentRenderer(<Pagination pageCount={10} selectedPageIndex={0} onSelectedPageIndexChange={onChange} />),
    );
    const buttons = queryAllByRole('button');
    expect(buttons.length).toEqual(9);
    fireEvent.click(buttons[4]);
    expect(onChange).toBeCalledTimes(1);
    expect(onChange).toBeCalledWith(3);
  });
});
