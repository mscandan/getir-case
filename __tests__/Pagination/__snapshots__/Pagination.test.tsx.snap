// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Tests should match snapshot 1`] = `
<ul
  className="sc-jIZahH bkLTRs"
>
  <li
    className="previous disabled"
  >
    <a
      aria-disabled="true"
      aria-label="Previous page"
      className=" "
      onClick={[Function]}
      onKeyPress={[Function]}
      rel="prev"
      role="button"
      tabIndex="-1"
    >
      <div
        className="sc-ikZpkk btchlR"
      >
        <div
          className="sc-himrzO hAGDXI"
        >
          Prev
        </div>
      </div>
    </a>
  </li>
  <li
    className="rp-page selected"
  >
    <a
      aria-current="page"
      aria-label="Page 1 is your current page"
      className="rp-link"
      onClick={[Function]}
      onKeyPress={[Function]}
      rel="active"
      role="button"
      tabIndex="-1"
    >
      1
    </a>
  </li>
  <li
    className="rp-page"
  >
    <a
      aria-current={null}
      aria-label="Page 2"
      className="rp-link"
      onClick={[Function]}
      onKeyPress={[Function]}
      rel="next"
      role="button"
      tabIndex="0"
    >
      2
    </a>
  </li>
  <li
    className="rp-page"
  >
    <a
      aria-current={null}
      aria-label="Page 3"
      className="rp-link"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
    >
      3
    </a>
  </li>
  <li
    className="rp-break"
  >
    <a
      className="rp-break-link"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
    />
  </li>
  <li
    className="rp-page"
  >
    <a
      aria-current={null}
      aria-label="Page 8"
      className="rp-link"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
    >
      8
    </a>
  </li>
  <li
    className="rp-page"
  >
    <a
      aria-current={null}
      aria-label="Page 9"
      className="rp-link"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
    >
      9
    </a>
  </li>
  <li
    className="rp-page"
  >
    <a
      aria-current={null}
      aria-label="Page 10"
      className="rp-link"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
    >
      10
    </a>
  </li>
  <li
    className="next"
  >
    <a
      aria-disabled="false"
      aria-label="Next page"
      className=""
      onClick={[Function]}
      onKeyPress={[Function]}
      rel="next"
      role="button"
      tabIndex="0"
    >
      <div
        className="sc-ikZpkk btchlR"
      >
        <div
          className="sc-himrzO hAGDXI"
        >
          Next
        </div>
      </div>
    </a>
  </li>
</ul>
`;
